File: src/lib/reflexes.ts
/* Reflexes (SUB-826) — the frontend's view of the rules engine.

   Types only, plus the two small pure helpers the settings section needs. The
   app never authors rules: they live in `.vault/reflexes.json` and are edited
   there, so there is nothing here that builds or writes one. What the pane does
   own is the one decision the file cannot make for itself — whether rules may
   run on THIS device for THIS vault — and reporting what they have been doing.

   Rust owns everything else (parsing, validation, execution, receipts); these
   shapes mirror `commands::reflexes` and `reflexes::run::Receipt`. */

/** One rule as the pane shows it: the file's declaration plus the runtime's
    memory of it. `enabled`/`dryRun` come from the file; `autoPaused`,
    `lastFired`, `lastError` and `suppressed` are runtime state that is never
    written back into the file. */
export type ReflexRule = {
  id: string;
  event: string;
  path: string | null;
  actions: string[];
  enabled: boolean;
  dryRun: boolean;
  autoPaused: boolean;
  lastFired: string | null;
  lastError: string | null;
  suppressed: number;
};

/** A rule that did not pass validation, kept with its reason so a broken rule
    is visibly broken instead of quietly missing. */
export type ReflexInvalid = { id: string; error: string };

export type ReflexStatus = {
  /** Has this device ever enabled reflexes for this vault? */
  enabled: boolean;
  /** Enabled, then paused by the user. */
  paused: boolean;
  enabledAt: string | null;
  /** The file's own top-level kill switch — a different switch from `paused`. */
  filePaused: boolean;
  hasFile: boolean;
  error: string | null;
  rules: ReflexRule[];
  invalid: ReflexInvalid[];
};

export type ReflexReceipt = {
  at: string;
  rule: string;
  event: string;
  subject: string;
  actions: string[];
  outcome: string;
  dryRun?: boolean;
};

/** What the section as a whole is: which of these four states it is in decides
    whether the pane leads with an enable card, a rule list, or nothing at all.

    - `absent`  — no rules file; nothing to say, so the section hides itself
    - `offer`   — a rules file exists and this device has never armed it: the
                  amendment's first-run state, shown as paused behind one switch
    - `paused`  — armed, then paused (by the user or by the file)
    - `live`    — armed and running */
export type ReflexSectionState = "absent" | "offer" | "paused" | "live";

export function sectionState(s: ReflexStatus | null): ReflexSectionState {
  // an unloadable file still counts as present: "your rules are broken" is
  // exactly the thing a hidden section would swallow
  if (!s || (!s.hasFile && !s.error)) return "absent";
  if (!s.enabled) return "offer";
  if (s.paused || s.filePaused) return "paused";
  return "live";
}

/** One line summarising a rule for the pane: last outcome first, because that
    is what someone opening this section wants to know. Deliberately not a
    status word alone — "never fired" and "fired, then failed" are different
    situations and a single badge would blur them. */
export function ruleSummary(r: ReflexRule): string {
  if (r.lastError) return `last error: ${r.lastError}`;
  if (r.lastFired) return `last fired ${r.lastFired}`;
  return "not fired yet";
}

File: src-tauri/src/commands/reflexes.rs
//! Reflexes: the one-time per-vault enable switch, what the rules file says,
//! and what the rules have actually been doing (SUB-826).
//!
//! Five commands and no sixth: nothing here writes `reflexes.json`. Rules are
//! authored in the file — the app reads them, reports them, and decides whether
//! they may run on this device. Keeping rule authoring out of the app is what
//! keeps the file the single source of truth, the same way keeping bundle
//! install out of the app is what makes the kinds enable card meaningful
//! (`commands::kinds`).
//!
//! Consent is written to `config.json` in the OS app-config dir, keyed by
//! canonical vault path. See `crate::reflexes::consent` for why it cannot live
//! in the vault.

use crate::reflexes::{self, consent, run, InvalidRule};
use crate::{AppState, OnboardingState};
use serde::Serialize;
use tauri::State;

/// One rule as the settings pane shows it: what the file says, plus what the
/// runtime remembers about it.
#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
pub(crate) struct RuleView {
    pub id: String,
    pub event: String,
    pub path: Option<String>,
    pub actions: Vec<String>,
    /// The file's own switch.
    pub enabled: bool,
    pub dry_run: bool,
    /// Paused by the circuit breaker after repeated failures — runtime state,
    /// never written back to the file.
    pub auto_paused: bool,
    pub last_fired: Option<String>,
    pub last_error: Option<String>,
    pub suppressed: u32,
}

/// Everything the Reflexes settings section renders in one round trip, so the
/// list can never be one call stale against the switch that governs it.
#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
pub(crate) struct ReflexStatus {
    /// Has this device ever enabled reflexes for this vault?
    pub enabled: bool,
    /// Enabled, then paused by the user.
    pub paused: bool,
    pub enabled_at: Option<String>,
    /// The file's own top-level kill switch — a different switch from `paused`.
    pub file_paused: bool,
    /// Is there a rules file at all?
    pub has_file: bool,
    /// Why the file didn't load, when it didn't.
    pub error: Option<String>,
    pub rules: Vec<RuleView>,
    pub invalid: Vec<InvalidRule>,
}

#[tauri::command]
pub(crate) fn reflexes_status(
    state: State<AppState>,
    onboarding: State<OnboardingState>,
    reflexes: State<reflexes::ReflexState>,
) -> ReflexStatus {
    let root = state.0.lock().unwrap().root.clone();
    let status = consent::status(&onboarding.config_dir, &root);
    let loaded = reflexes.0.lock().unwrap();
    let rules = loaded
        .reflexes
        .rules
        .iter()
        .map(|r| {
            let st = loaded.runtime.state(&r.id);
            RuleView {
                id: r.id.clone(),
                event: r.event.as_str().to_string(),
                path: r.path.clone(),
                actions: r.actions.iter().map(|a| a.verb().to_string()).collect(),
                enabled: r.enabled,
                dry_run: r.dry_run,
                auto_paused: st.map(|s| s.auto_paused).unwrap_or(false),
                last_fired: st.and_then(|s| s.last_fired.clone()),
                last_error: st.and_then(|s| s.last_error.clone()),
                suppressed: st.map(|s| s.suppressed).unwrap_or(0),
            }
        })
        .collect();
    ReflexStatus {
        enabled: status.enabled,
        paused: status.paused,
        enabled_at: status.enabled_at,
        file_paused: loaded.reflexes.paused,
        has_file: root.join(reflexes::CONFIG_REL_PATH).is_file(),
        error: loaded.error.clone(),
        rules,
        invalid: loaded.reflexes.invalid.clone(),
    }
}

/// Arm reflexes for this vault on this device. One switch for the whole
/// feature: after this, rule edits are just edits (the closed verb set is what
/// makes that safe — see `crate::reflexes::consent`).
#[tauri::command]
pub(crate) fn reflexes_enable(
    state: State<AppState>,
    onboarding: State<OnboardingState>,
) -> Result<(), String> {
    if cfg!(target_os = "ios") {
        return Err("reflexes run on desktop only".into());
    }
    let root = state.0.lock().unwrap().root.clone();
    consent::enable(&onboarding.config_dir, &root)
}

/// Stop for now, keeping the decision. Unlike `reflexes_disable` this does not
/// return the vault to its never-armed state, so re-arming is one click and not
/// a fresh grant of trust.
#[tauri::command]
pub(crate) fn reflexes_set_paused(
    state: State<AppState>,
    onboarding: State<OnboardingState>,
    paused: bool,
) -> Result<(), String> {
    let root = state.0.lock().unwrap().root.clone();
    consent::set_paused(&onboarding.config_dir, &root, paused)
}

/// Withdraw the enable entirely: back to the first-run state, rules shown but
/// never run.
#[tauri::command]
pub(crate) fn reflexes_disable(
    state: State<AppState>,
    onboarding: State<OnboardingState>,
) -> Result<(), String> {
    let root = state.0.lock().unwrap().root.clone();
    consent::disable(&onboarding.config_dir, &root)
}

/// The receipts, newest first. Read straight off `.vault/reflexes-log.json`,
/// which is the record — the runtime keeps no second copy to drift from it.
#[tauri::command]
pub(crate) fn reflexes_receipts(state: State<AppState>) -> Vec<run::Receipt> {
    let root = state.0.lock().unwrap().root.clone();
    let mut out = run::read_log(&root);
    out.reverse();
    out
}
